Assignment 5

agaga23 opened the pull request.

agaga23 committed:
Assignment 5

nuwanda57 commented on a5.cpp:
```diff
@@ -0,0 +1,107 @@
+#include<bits/stdc++.h>
+using namespace std;
```
Maybe it would make the code more readable if you add empty strings here to devide headers from class realisation.

agaga23 replied:
I don't use empty lines too much, because I like to see as much code as possible during writing.
But you are right :)

nuwanda57 commented on a5.cpp:
```diff
@@ -0,0 +1,107 @@
+		connectedWith.resize ( sizeOfAlphabet + 1, 0 );
+		
+		for ( int i = 0; i < dictionary.size(); i ++ ){
+			for ( int j = i + 1; j < dictionary.size(); j ++ ){
```
I think it will be enough just to compare nearby words. For example:
Dictionary: ART, RAT, CAT, CAR.
Comparing ART with RAT, you know that A<R;
Comparing RAT with CAT, you know that R<C;
So you don't need to compare ART with CAT because you already know that A<C (A<R<C).

As a result the complexity will be linear instead of square.

ssancho2 replied:
Agree.

agaga23 replied:
Good point, thank you! I haven't realised it before.

nuwanda57 commented on a5.cpp:
```diff
@@ -0,0 +1,107 @@
+		for ( int i = 0; i < minLength; i ++ )
+			if ( a[i] != b[i] ) return i;
+		
+		return minLength;
```
Maybe it would be better to return something like -1, and instead of 
if ( differentPos < dictionary[i].size() && differentPos < dictionary[j].size() ) in line 72
write
if (differentPos != -1)
I think it would be shorter and better for understanding.
Or maybe not:)

ssancho2 replied:
Agree

agaga23 replied:
Fixed

nuwanda57 commented on a5.cpp:
```diff
@@ -0,0 +1,107 @@
+using namespace std;
+class Alphabet{
+	map < char, int > allLetters;
+	vector < char > numberedLetters;
```
May be use the ASCII table instead of allLetters and numberedLetters?

agaga23 replied:
Good idea :) I didn't want to make a big array in constructor, but actually 200 is not too big.

nuwanda57 commented on a5.cpp:
```diff
@@ -0,0 +1,107 @@
+		}
+		return topSorting();
+	}
+};
```
I think it would be better to add blank lines here too

agaga23 replied:
Ok, added

nuwanda57 commented on a5.cpp:
```diff
@@ -0,0 +1,107 @@
+		if ( i + 1 < alphabet.size() ) cout << ", ";
+	}
+	cout << "]\n";
+}
```
And also here:)

agaga23 replied:
Ok

nuwanda57 commented on a5.cpp:
```diff
@@ -0,0 +1,107 @@
+vector<char> findTheAlphabet ( vector<string>& dictionary ){
+	Alphabet alpha;
+	return alpha.findTheAlphabet ( dictionary );
+}
```
And before the main function

agaga23 replied:
Ok, so just after every function? I added even some more :)

Thank you @julia57 for all your comments!

nuwanda57 reviewed (commented):
Good solution! Just some comments.

ssancho2 commented on a5.cpp:
```diff
@@ -0,0 +1,107 @@
+	}
+	int findAllLetters ( vector<string>& dictionary ){
+		int counter = 0;
+		for ( int i = 0; i < dictionary.size(); i ++ ){
```
Consider using `for (const string &word : dictionary) { ... }` , I think it makes code more readable, but it may be a mater of taste.

agaga23 replied:
Agree, changed

ssancho2 commented on a5.cpp:
```diff
@@ -0,0 +1,107 @@
+		for ( int i = 0; i < dictionary.size(); i ++ ){
+	
+			string word = dictionary[i];
+			for ( int j = 0; j < word.length(); j ++ ) {
```
Same here, you can do `for (char letter : word) { ... } `

agaga23 replied:
Changed

ssancho2 commented on a5.cpp:
```diff
@@ -0,0 +1,107 @@
+	vector < char > numberedLetters;
+	
+	vector < vector < int > > orderGraph;
+	vector < int > connectedWith;
```
I appreciate it's faster this way, and no need to change it, but usually it's better to have code that's more readable. In this case, you can have `map<char, set<char>> orderGraph` and `map<char, int> connectedWith` and let the compiler worry about indexing with chars :)
If you do want to do the mapping, better to have a class to encapsulate that, with methods like `addChar`, `mapCharToInt` and `mapIntToChar`

agaga23 replied:
Ok, I changed it as Julia suggested - without map, using ASCII. I think it is more readable now :)

ssancho2 replied:
Fair enough, just a word of warning that char can have 256 possible values, but it can be signed or unsigned depending on the compiler/platform. This is the kind of latent bug that make us says 'but it works on my computer!' :)

agaga23 replied:
Good to know. Changed

ssancho2 replied:
Just to be clear, the problem is indexing orderGraph & connectedWith with chars, you expect them to be in the 0 to 255 range, but they may be on the -127 to 128 range. As mentioned earlier, one solution is to change from `vector<T>` to `map<char, T>`, which may be a bit slower. Another one is to use something like `orderGraph[255 & (int) c]` when accessing the array, which is a bit messy.

ssancho2 commented on a5.cpp:
```diff
@@ -0,0 +1,107 @@
+					char letter1 = dictionary[i][differentPos], letter2 = dictionary[j][differentPos];
+					
+					int nrOfLetter1 = allLetters[letter1], nrOfLetter2 = allLetters[letter2];
+					cerr << letter1 << " < " << letter2 << "\n";
```
Please remove debugging statements :)

agaga23 replied:
Removed

ssancho2 commented on a5.cpp:
```diff
@@ -0,0 +1,107 @@
+	}
+	cout << "]\n";
+}
+vector<char> findTheAlphabet ( vector<string>& dictionary ){
```
Please make this parameter `const` (and most others that you are not really changing).

agaga23 replied:
Fixed

Thank you @ssancho2 for your review :)

ssancho2 commented on a5.cpp:
```diff
@@ -0,0 +1,107 @@
+#include<bits/stdc++.h>
```
A few comments won't hurt :)

agaga23 replied:
I forgot, sorry. Added

ssancho2 reviewed (commented):
Some comments, looks good!

agaga23 committed:
After comments

ssancho2 commented on a5.cpp:
```diff
@@ -1,107 +1,108 @@
 		
-		orderGraph.resize ( sizeOfAlphabet + 1 );
-		connectedWith.resize ( sizeOfAlphabet + 1, 0 );
+		orderGraph.resize ( 200 );
```
Why not 256?

agaga23 replied:
I checked on wiki that the possible values are from 0 to 127, I haven't realised that it gives 128 possibilities :D 200 seemed to be safe and round. Now I see it was stupid approach.

ssancho2 commented on a5.cpp:
```diff
@@ -1,107 +1,108 @@
-		for ( int i = 0; i < dictionary.size(); i ++ ){
-			for ( int j = i + 1; j < dictionary.size(); j ++ ){
-				int differentPos = firstDifferentPosition ( dictionary[i], dictionary[j] );
+		for ( int i = 0; i < dictionary.size() - 1; i ++ ){
```
Very subtle bug here: dictionary.size() is unsigned, when is 0 and you subtract 1 what happens?

agaga23 replied:
Will it be 2^32 - 1?
I have never thought about this!! Thanks a lot!

ssancho2 commented on a5.cpp:
```diff
@@ -1,107 +1,108 @@
-				}
+				const char letter1 = dictionary[i][differentPos], letter2 = dictionary[i + 1][differentPos];
+				
+				orderGraph[letter1].push_back ( letter2 ); //connection: letter1 should be in the alphabet before letter2
```
Minor comment on performance, you may have repeated letters on the vectors. If you fix this be careful to only increase `connectedWith` once.

agaga23 replied:
Fixed

ssancho2 commented on a5.cpp:
```diff
@@ -1,107 +1,108 @@
+	
 	dict = { "z", "yx", "yz", "a" };
 	printAlphabet ( findTheAlphabet ( dict ) );
+	
```
The empty dictionary is a interesting test case :), same with a one word dictionary.

agaga23 replied:
Added, it work for both tests :)

ssancho2 replied:
Good :)

ssancho2 reviewed (commented):
A few more things

agaga23 committed:
After comments2

alphabet size changed to 256
bug with .size() - 1 fixed
added tests
improved performace, deleted multiple edges

agaga23 committed:
changed vectors into maps of chars

Oh, I see the difference. Could I also change numeration of chars +127 to their values to get nonnegative numbers ? I know it would be much less readable solution but I wanted to ensure if it was correct.

agaga23 commented:
I changed vector to map. I am not sure if I understood your second solution. Wouldn't your second solution (255 & (int) c) give abs(c) ?

ssancho2 commented:
`c + 128` works for signed chars, since c is promoted to integer and then the addition takes place, but for unsigned chars it maps to the range 128 - 382, which is not what you want.
`(255 & (int) c)` always gives you a number in the range 0 - 255, it's obvious for unsigned chars, since both the `(int)` and `255 &` are no-ops. When c is signed, numbers 0-127 are mapped the 0-127, again, both are no-ops. When c is -1 for example, it's converted to -1 (integer), which since is https://en.wikipedia.org/wiki/Two%27s_complement is all binary 1s, then the `255 &` maps it to 255, -2 is mapped to 254 and so on until -128 that is mapped to 128.

agaga23 commented:
Now I think I understand it. Thank you :)
